### ISSUE TITLE
build/cmake: consider Hunter (vcpkg, ?) for bundled dependencies

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#8446]
[This thread](https://news.ycombinator.com/item?id=17163194) discusses CMake state-of-the-art. 

Currently we use `ExternalProject_Add` combined with our own recipes for building external projects. This obviously is painful to maintain because if an upstream project changes it may break our bundled-deps build when we bump our dependency version of that project.

Some recommend [hunter](https://github.com/cpp-pm/hunter) for pulling in external projects with CMake.

>  Hunter aims to be a full package manager, so it strives to build a project and all of its depedendencies recursively. This is fundamentally different than just adding one external project and ignoring the depedencies (what ExternalProject_Add does). The downside is this forces Hunter to maintain many forks of original repos just to keep track of the dependency information. 

[Quick introduction to Hunter](https://geokon-gh.github.io/hunterintro.html):

> * **Dependencies are built at configuration time:** Hunter will download the source and build all your dependencies during configuration time.
> * **Dependencies are built independently:** Each dependency is configured/built/installed independently and the only thing they inherit from the parent project is the toolchain file you passed in at the command line.
> * **Dependencies share their own dependencies:** If two dependencies need zlib then Hunter will install just one zlib and link both dependencies to it

# Evaluation

### conan

- conan is missing many of our deps (libtermkey, libuv at least)
- fast to install on Arch Linux (Python package)

### vcpkg 

- vcpkg appears to provide more deps, but still also misses libtermkey, libvterm at least.
- slow to install (only vcpkg-git in Arch, requires to be built)

### hunter

- hunter does not need to be installed really (via CMake)
- [min cmake version](https://hunter.readthedocs.io/en/latest/quick-start/cmake.html)
- also misses a lot of our deps, including LuaJIT: https://hunter.readthedocs.io/en/latest/packages.html

### nix

- ?

**Comment by jszakmeister** [comment]
@justinmk I've not personally played with hunter at all--in my work environment, it's simply a no-go.  I do wonder about messing up people's environments though.  Having the builds all centralized into one location is nice in some respects and limiting in others, such as when we have to apply patches to workaround issues or dealing with versions mismatches (we require some version, but it's been updated out from under us, etc.).  I also have no clue how this affects packaging either--it's not clear if it'll use a system resource when available or not.

Hopefully someone with more experience here can chime in!

**Comment by ruslo** [comment]
Hi,

> limiting in others, such as when we have to apply patches to workaround issues or dealing with versions mismatches (we require some version, but it's been updated out from under us, etc.).

How you're dealing with it currently? If you have a fork with workarounds you can make a custom release there and use it in Hunter. There are a lot of other options as well, it really depends on your needs.

> I also have no clue how this affects packaging either--it's not clear if it'll use a system resource when available or not.

What do you mean by packaging?

Probably it will be easier to focus discussion on real examples. E.g. if we are talking about `libuv` then using Hunter instead of custom ExternalProject_Add will mean removing https://github.com/neovim/neovim/blob/418abfc9d06923e96f1317419fe83ed4e6d67c1d/third-party/cmake/BuildLibuv.cmake completely and substituting https://github.com/neovim/neovim/blob/418abfc9d06923e96f1317419fe83ed4e6d67c1d/CMakeLists.txt#L321-L322 with:

```cmake
hunter_add_package(libuv)
find_package(libuv CONFIG REQUIRED)
target_link_libraries(... libuv::uv)
```

See:
* https://docs.hunter.sh/en/latest/packages/pkg/libuv.html

**Comment by jamessan** [comment]
@ruslo How can we remove BuildLibuv.cmake completely, when we have a Windows-specific patch for libuv?

https://github.com/neovim/neovim/blob/418abfc9d06923e96f1317419fe83ed4e6d67c1d/third-party/cmake/BuildLibuv.cmake#L45-L48

We have similar patches for other dependencies and have specifically used the patching functionality instead of creating our own forks for each project.

**Comment by ruslo** [comment]
Fork is a preferred way to deal with patches in Hunter:
* https://docs.hunter.sh/en/latest/faq/why-do-we-need-forks.html

**Comment by jszakmeister** [comment]
> What do you mean by packaging?

I mean does it affect our upstream maintainers who take care of packaging Neovim for various platforms?  Debian, Ubuntu, Fedora, etc. get awfully picky about their environments and we don't want to go in a direction that's going to make the job harder (it's hard enough as it is!). :-)

> hunter_add_package(libuv)
> find_package(libuv CONFIG REQUIRED)
> target_link_libraries(... libuv::uv)

This change seems more invasive than it looks.  Currently, we have the dependencies being built separately from the main application.  This allows packagers to skip the building of the dependencies and just use CMake to drive building the app directly (they'll ensure the system libraries are there and available).

Does going this route require that we use a Hunter-supplied version versus a system-supplied one?  That's probably the biggest issue that I can see.  Many of the packaging environments don't want you to download from places on the Internet--they want you to only use packaged libraries.  If Hunter requires that we use their version, that's a bit of a problem.  We could wrap `hunter_add_package()` in a conditional and maybe that helps as long as `libuv::uv` isn't a Hunter-specific name.

Also, forking repos things a bit tougher to manage our patches since they're out of the Neovim tree.

**Comment by ruslo** [comment]
> I mean does it affect our upstream maintainers who take care of packaging Neovim for various platforms?

If on those platforms you want to use system libraries you can turn Hunter OFF:
* https://docs.hunter.sh/en/latest/overview/compatibility.html

**Comment by jszakmeister** [comment]
> If on those platforms you want to use system libraries you can turn Hunter OFF:

Awesome!

**Comment by arecarn** [comment]
Just to throw another hat in the ring. There is https://conan.io/ which also supports CMake and is gaining a lot of traction in the C++ community.

**Comment by ras0219-msft** [comment]
[Dev on vcpkg]

I think there are three key features of vcpkg that make it really appealing for a large application like neovim:

1. Vcpkg is designed from the ground up to be truly embeddable -- it can function like a drop-in replacement for checking in all your dependency's code, without having to actually check in their code. We accomplish this with SHA512 hashes for all downloads, ensuring you truly get bit-for-bit identical inputs. If desired, the entire tool (and all dependency scripts) can be embedded in the neovim project via `git subtree` and there will be _no_ dependencies on the upstream vcpkg project.
2. We take great pains to ensure compatibility with builtin CMake modules (and therefore with the system package managers like brew/apt). It is a goal of vcpkg to support projects that want to have the _option_ of bundling dependencies, but don't want to be locked away from supporting system package managers like brew/apt/pacman/etc.
3. Native support for exporting prebuilt "dependency archives". Running `vcpkg export zlib --zip` produces a completely freestanding binary archive that can consumed exactly the same as a full vcpkg instance, providing the precise dependencies that are used in upstream CI builds and allowing developers to get started working on _neovim itself_ really quickly! This binary archive is host agnostic and could be attached to GitHub releases, your own website, bintrayer, or elsewhere.

I'd love to chat more about what features you'd want out of a package manager via vcpkg@microsoft.com!

**Comment by ras0219-msft** [comment]
>the other major requirement for us is cross-platform support.

In case you haven't seen: https://blogs.msdn.microsoft.com/vcblog/2018/04/24/announcing-a-single-c-library-manager-for-linux-macos-and-windows-vcpkg/. I also know people successfully using us on FreeBSD and using us to target iOS, Android, and Emscripten. Essentially, anything you can get a toolchain file for can be targeted by vcpkg!

>To support this, I can imagine a "snapshot" feature … Does this make sense?

Definitely, we've been strongly looking at adding a feature like that already; it's great to see another +1 :)

>Maybe the answer is that vcpkg hosts forks/mirrors of all the dependencies

We don't generally rehost sources[1], we instead host patches that are applied on top of the original sources (fetched from the upstream locations).

Today, we cache all downloads in the `downloads\` directory under the vcpkg system. This means that a crude "snapshot" could be taken today by building your entire dependency graph once, then archive the entire `downloads\` directory. This will include all the source archives, along with any needed runtime tools (msys2, perl, cmake, ninja, nasm, etc). It is our goal that no network access should be required during a rebuild and we actively patch out attempts by inner buildsystems to access the network.

An even simpler "brute force" approach would be to _check in_ the downloads directory in-place to a vcpkg fork, which would make bisecting trivial :).

[1] There are a small number of `Microsoft/X` forks we pull from, however these aren't directly tied with vcpkg

**Comment by blueyed** [comment]
Some remarks while looking into this:

### conan

- conan does not contain quite some of our deps (libtermkey, libuv at least)
- fast to install on Arch Linux (Python package)

### vcpkg 

- vcpkg appears to have a few more deps, but still also misses libtermkey, libvterm at least.
- slow to install (only vcpkg-git in Arch, requires to be built)

### hunter

- hunter does not need to be installed really (via CMake)
- requires CMake 3.1.2 (have not checked the others)
- also misses a lot of our deps, including LuaJIT (https://docs.hunter.sh/en/latest/genindex.html)

We should probably have a table in the top comment summarizing this, and having it more structured for comparison.

> meson

That would be more involved, and is out of scope I guess.

**Comment by ras0219-msft** [comment]
>slow to install (only vcpkg-git in Arch, requires to be built)

We recommend pulling vcpkg directly in your project and not relying on a copy from the system package manager -- this is probably easiest with git submodules, but other mechanisms work fine too!

**Comment by eli-schwartz** [comment]
> I know I'm the one that suggested vcpkg in the first place, but I feel bad if I don't also mention meson, which does something similar with regards to dependencies: http://mesonbuild.com/Wrap-dependency-system-manual.html
> 
> (meson has a hard dependency on python 3.5+ for anyone wanting to build from source, which is why we didn't ultimately go with it for fish. But what's considered acceptable for a shell may not be the same for an editor, and neovim certainly has a hell of a lot more dependencies than fish does.)

The situation has evolved, by the way: https://mesonbuild.com/FAQ.html#but-i-really-want-a-version-of-meson-that-doesnt-use-python

There is now a c99 reimplementation of Meson called [muon](https://muon.build/) with a statically linked single-file version weighing in at 4.5 mb; it can be bootstrapped with `cc` and `sh`  and optionally depends on pkgconf, libcurl, and libarchive. It has pretty good support for the core functionality (C/C++, custom_target, etc.) and limited modules support (the pkgconfig and fs modules) and is a viable solution for building many projects -- hopefully more, as people implement the remaining functionality.

You don't necessarily need to use the python-based reference implementation any more, so, depending on your use cases, that may make Meson become appealing again.

**Comment by dundargoc** [comment]
I'm removing CPM as a candidate for the following reasons:

- It is basically just a fancy wrapper over FetchContent, so it's not nearly in
  the same ball game with the other suggestions.

- Because it's a wrapper over FetchContent, it suffers from its main drawback:
  doesn't work for dependencies that doesn't use cmake. Unfortunately for us we
  have many of those.

- We already have many of its benefits, such as offline builds with
  `USE_EXISTING_SRC_DIR=ON`.

**Comment by dundargoc** [comment]
I don't think it's as simple as you make it out to be. None of the following can be helped with Makefile.

- Makefiles are non-standard on Windows. Everyone uses cmake, and no one uses makefile.

- We'd need virtualization for arm mac up and running to get releases for both macs. At least I didn't see a way to dual compile.

- We basically need to add vcpkg as a submodule since the repo will be always needed. Whether this be with `git submodule` or some other way I don't know. Because of this most of CI/build system will require a significant overhaul. Aka this will probably require work and not be a cute drop-in replacement.

- Losing support for a bunch of non-standard systems. I personally don't really care but I know some people care about this a lot. Apparently people don't like it when you say "lmao who tf uses BSD in $CURRENT_YEAR?".

- Tangential note: this is a HUGE vendor lock-in, and probably something we shouldn't do lightly.

**Comment by dundargoc** [comment]
I attempted to integrate vcpkg in https://github.com/neovim/neovim/pull/26431. I'll be frank: I don't think we can use vcpkg in its current state. Essentially the solution to every roadblock has been a non-solution: implement it yourself. I could not find a satisfactory answer to any of the following problems:

- How to specify alternative dependencies. There seems to be no way to specify if we want to use lua or luajit when we download e.g. lpeg. This would've been fine if it weren't for...

- No builtin way to not install dependencies. You will get each dependency and you will like it. This leads to...

- Multiple versions of the same dependency, which fails the build. There was no builtin way to resolve this except for "specify everything yourself", which sounds more work than what we're currently doing.

- No way to point to a specific commit of a dependency. This is basically a showstopper as the latest luajit commit we can use is from 2023-01-04.

- Basically forced to add vcpkg repo as a submodule in order to specify versions of dependencies. Wat.

Doing basically anything results in "write your own portfile/registry/etc" which is defeats the purpose of a package manager in the first place.

**Comment by justinmk** [comment]
I expect zig to obviate the need for this.
